From 5ac1de1b200cbf15cac9f91b5c1b1075d5567003 Mon Sep 17 00:00:00 2001
From: Victor Fuentes <victor@vlinkz.dev>
Date: Wed, 29 Jul 2026 01:16:39 -0700
Subject: [PATCH 099/138] drm/msm/dsi/phy: do not start the 7nm PLL before a
 rate has been set

Starting it from its reset configuration fails to lock when the
bootloader handed over a running display.

Signed-off-by: Victor Fuentes <victor@vlinkz.dev>
---
 drivers/gpu/drm/msm/dsi/phy/dsi_phy_7nm.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/msm/dsi/phy/dsi_phy_7nm.c b/drivers/gpu/drm/msm/dsi/phy/dsi_phy_7nm.c
index 5d805a797abd..5750066a645d 100644
--- a/drivers/gpu/drm/msm/dsi/phy/dsi_phy_7nm.c
+++ b/drivers/gpu/drm/msm/dsi/phy/dsi_phy_7nm.c
@@ -96,6 +96,7 @@ struct dsi_pll_7nm {
 	 */
 	spinlock_t pll_enable_lock;
 	int pll_enable_cnt;
+	bool vco_configured;
 
 	struct pll_7nm_cached_state cached_state;
 
@@ -372,6 +373,8 @@ static int dsi_pll_7nm_vco_set_rate(struct clk_hw *hw, unsigned long rate,
 	/* flush, ensure all register writes are done*/
 	wmb();
 
+	pll_7nm->vco_configured = true;
+
 	return 0;
 }
 
@@ -496,6 +499,9 @@ static int dsi_pll_7nm_vco_prepare(struct clk_hw *hw)
 	struct dsi_pll_7nm *pll_7nm = to_pll_7nm(hw);
 	int rc;
 
+	if (!pll_7nm->vco_configured)
+		return 0;
+
 	dsi_pll_enable_pll_bias(pll_7nm);
 	if (pll_7nm->slave)
 		dsi_pll_enable_pll_bias(pll_7nm->slave);
@@ -549,6 +555,9 @@ static void dsi_pll_7nm_vco_unprepare(struct clk_hw *hw)
 {
 	struct dsi_pll_7nm *pll_7nm = to_pll_7nm(hw);
 
+	if (!pll_7nm->phy->pll_on)
+		return;
+
 	/*
 	 * To avoid any stray glitches while abruptly powering down the PLL
 	 * make sure to gate the clock using the clock enable bit before
